File: go/101-house/house.go
package house

import (
	"fmt"
	"strings"
)

var verse = []struct{ animal, action string }{
	{"house that Jack built", "lay in"},
	{"malt", "ate"},
	{"rat", "killed"},
	{"cat", "worried"},
	{"dog", "tossed"},
	{"cow with the crumpled horn", "milked"},
	{"maiden all forlorn", "kissed"},
	{"man all tattered and torn", "married"},
	{"priest all shaven and shorn", "woke"},
	{"rooster that crowed in the morn", "kept"},
	{"farmer sowing his corn", "belonged to"},
	{"horse and the hound and the horn", ""},
}

func Verse(v int) string {
	var output string

	for i := v; i >= 1; i-- {
		if i == v {
			output += fmt.Sprintf("This is the %s", verse[v-1].animal)
		} else {
			output += fmt.Sprintf("\nthat %s the %s", verse[i-1].action, verse[i-1].animal)
		}
	}

	return output + "."
}

func Song() string {
	var verses []string

	for i := 1; i < 13; i++ {
		verses = append(verses, Verse(i))
	}

	return strings.Join(verses, "\n\n")
}

File: ruby/097-house/house.rb
class House
  PIECES = [
    ['the malt', 'that lay in'],
    ['the rat', 'that ate'],
    ['the cat', 'that killed'],
    ['the dog', 'that worried'],
    ['the cow with the crumpled horn', 'that tossed'],
    ['the maiden all forlorn', 'that milked'],
    ['the man all tattered and torn', 'that kissed'],
    ['the priest all shaven and shorn', 'that married'],
    ['the rooster that crowed in the morn', 'that woke'],
    ['the farmer sowing his corn', 'that kept'],
    ['the horse and the hound and the horn', 'that belonged to']
  ].freeze

  def self.recite
    verses = ["This is the house that Jack built."]

    (0...PIECES.length).each do |i|
      current_verse_lines = ["This is #{PIECES[i][0]}"]
      (i.downto(0)).each do |j|
        line = if j == 0
                 "#{PIECES[j][1]} the house that Jack built."
               else
                 "#{PIECES[j][1]} #{PIECES[j - 1][0]}"
               end
        current_verse_lines << line
      end
      verses << current_verse_lines.join("\n")
    end

    verses.join("\n\n") + "\n"
  end
end